File: java/lista_dinamica/ListaDinamica.java
package lista_dinamica;

import main.Celula;

public class ListaDinamica {
	public Celula primeiraAuxiliar, ultima;
	private int qtd;

	public ListaDinamica() {
		primeiraAuxiliar = new Celula();
		ultima = primeiraAuxiliar;
		qtd = 0;
	}

	public boolean isListaVazia() {
		return ultima == primeiraAuxiliar;
	}

	public int getTamanhoDaLista() {
		return qtd;
	}

	public boolean inserirNoInicio(int item) {
		Celula auxiliar = new Celula(item);

		if (isListaVazia()) {
			ultima = auxiliar;
		}

		auxiliar.prox = primeiraAuxiliar.prox;
		primeiraAuxiliar.prox = auxiliar; 
		qtd++;
		auxiliar = null;

		return true;
	}

	public boolean inserirNoFim(int item) {
		Celula auxiliar = new Celula(item);

		if (isListaVazia()) {
			primeiraAuxiliar.prox = auxiliar;
		}

		ultima.prox = auxiliar;
		ultima = auxiliar;

		qtd++;
		auxiliar = null;

		return true;
	}

	public boolean inserirPos(int item, int pos) {
		if (pos < 0 || pos > qtd) {
			System.out.println("Falha ao inserir: posição de inserção inválida.");
			return false;
		}

		Celula auxiliar = new Celula(item);
		Celula tmp = primeiraAuxiliar;
		for (int cont = 0; cont < pos; tmp = tmp.prox, cont++);
		auxiliar.prox = tmp.prox;
		tmp.prox = auxiliar;
		qtd++;
		auxiliar = null;

		return true;
	}

	public int removerDoInicio() {
		int itemRemovido = Integer.MIN_VALUE;

		if (isListaVazia()) {
			System.out.println("Falha ao remover: lista vazia!");
			return itemRemovido;
		}

		Celula auxiliar = primeiraAuxiliar;
		primeiraAuxiliar = primeiraAuxiliar.prox;
		auxiliar.prox = null;
		auxiliar = null;
		itemRemovido = primeiraAuxiliar.item;
		qtd--;

		return itemRemovido;
	}

	public int removerDoFim() {
		int itemRemovido = Integer.MIN_VALUE;

		if (isListaVazia()) {
			System.out.println("Falha ao remover: lista vazia!");
			return itemRemovido;
		}

		Celula auxiliar = primeiraAuxiliar.prox;
		for (; auxiliar.prox != ultima; auxiliar = auxiliar.prox);
		itemRemovido = ultima.item;
		auxiliar.prox = null;
		ultima = auxiliar;
		qtd--;
		auxiliar = null;

		return itemRemovido;
	}

	public int removerPos(int pos) {
		int itemRemovido = Integer.MIN_VALUE;
		if (pos < 0 || pos >= qtd) {
			System.out.println("Falha ao remover: posição de remoção inválida.");
			return itemRemovido;
		} else if (pos == 0) {
			itemRemovido = removerDoInicio();
		} else if (pos == (qtd-1)) {
			itemRemovido = removerDoFim();
		} else {
			Celula auxiliar = primeiraAuxiliar;
			for (int cont = 0; cont < pos; auxiliar = auxiliar.prox, cont++);
			Celula aRemover = auxiliar.prox;
			auxiliar.prox = aRemover.prox;
			aRemover.prox = null;
			itemRemovido = aRemover.item;
			aRemover = null;
			auxiliar = null;
			qtd--;
		}

		return itemRemovido;
	}

	public int pesquisarItem(int item) {
		int pos = -1, cont = 0;
		for (Celula auxiliar = primeiraAuxiliar.prox; auxiliar != null && (pos == -1); auxiliar = auxiliar.prox, cont++) {
			if (auxiliar.item == item) {
				pos = cont;
			}
		}
		return pos;
	}

	public void mostrarLista() {
		System.out.print("[ ");
		for (Celula auxiliar = primeiraAuxiliar.prox; auxiliar != null; auxiliar = auxiliar.prox) {
			System.out.print(auxiliar.item+" ");
		}
		System.out.println("]");
	}
}

File: c/lista_dinamica/listaDinamica.c
#include "listaDinamica.h"
#include "../common/celula.h"

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

Lista * newLista() {
    Lista * lista = (Lista *) malloc(sizeof(Lista));

    lista->primeiro = newCelula();
    lista->ultimo = lista->primeiro;
    lista->qtd = 0;

    return lista;
}

bool isVazia(Lista * lista) {
	return lista->primeiro == lista->ultimo;
}

bool inserirNoInicio(Lista * lista, int item) {
    bool deuCerto = false;
    Celula * c = newCelula2(item);
    if (isVazia(lista)) {
        lista->ultimo = c;
    }
    c->prox = lista->primeiro->prox;
    lista->primeiro->prox = c;
    c = NULL;
    deuCerto = true;
    lista->qtd++;
    return deuCerto;
}

bool inserirNoFim(Lista * lista, int item) {
    bool deuCerto = false;
    Celula * c = newCelula2(item);
    if (isVazia(lista)) {
        lista->primeiro->prox = c;
    }
    lista->ultimo->prox = c;
    lista->ultimo = c;
    c = NULL;
    deuCerto = true;
    lista->qtd++;
    return deuCerto;
}

bool inserirNaPos(Lista * lista, int item, int pos) {
    bool deuCerto = false;
    if (pos >= 0 && pos <= lista->qtd) {
        if (pos == 0) {
            deuCerto = inserirNoInicio(lista, item);
        } else if (pos == lista->qtd) {
            deuCerto = inserirNoFim(lista, item);
        } else {
            Celula * c = newCelula2(item);
            Celula * tmp = lista->primeiro->prox;
            for (int cont = 1; cont < pos; cont++) {
                tmp = tmp->prox;
            }
            c->prox = tmp->prox;
            tmp->prox = c;
            tmp = NULL;
            c = NULL;
			lista->qtd++;
        }
    } else {
        printf("Posição de inserção inválida.\n");
    }
    return deuCerto;
}

int removerDoInicio(Lista * lista) {
    int removido = 0x7fffffff;
    if (isVazia(lista)) {
        printf("Lista vazia!\n");
    } else {
        Celula * c = lista->primeiro;
        lista->primeiro = lista->primeiro->prox;
        free(c);
        removido = lista->primeiro->item;
        c = NULL;
        lista->qtd--;
    }
    return removido;
}

int removerDaPos(Lista * lista, int pos) {
    int removido = 0x7fffffff;
    if (isVazia(lista)) {
        printf("Lista vazia!\n");
    } else {
        if (pos >= 0 && pos < lista->qtd) {
            Celula * c = lista->primeiro;
            for (int cont = 0; cont < pos; cont++) {
                c = c->prox;
            }
            Celula * tmp = c->prox;
            c->prox = tmp->prox;
            removido = tmp->item;
            free(tmp);
            tmp = NULL;
            c = NULL;
            lista->qtd--;
        } else {
            printf("Posição de remoção inválida.\n");
        }
    }
    return removido;
}

int removerDoFim(Lista * lista) {
    int removido = 0x7fffffff;
    if (isVazia(lista)) {
        printf("Lista vazia!\n");
    } else {
        Celula * c = lista->primeiro;
        while (c->prox != lista->ultimo) {
            c = c->prox;
        }
        Celula * tmp = c->prox;
        c->prox = NULL;
        lista->ultimo = c;
        removido = tmp->item;
        free(tmp);
        tmp = NULL;
        c = NULL;
        lista->qtd--;
    }
    return removido;
}

bool pesquisarItemNaLista(Lista * lista, int item) {
	bool encontrado = false;
	for (Celula * tmp = lista->primeiro->prox; tmp != NULL && !encontrado; tmp = tmp->prox) {
		if (tmp->item == item) encontrado = true;
	}
	return encontrado;
}

int getPosicaoDoItemNaLista(Lista * lista, int item) {
	int pos = -1, cont = 0;
	for (Celula * tmp = lista->primeiro->prox; tmp != NULL && (pos == -1); tmp = tmp->prox) {
		if (tmp->item == item) pos = cont;
	}
	return pos;
}

void mostrarItensDaLista(Lista * lista) {
    if (lista->primeiro == lista->ultimo) {
        printf("Lista vazia!\n");
    } else {
        printf("[ ");
        for (Celula * c = lista->primeiro->prox; c != NULL; c = c->prox)
            printf("%d ", c->item);
        printf("]\n");
    }
}

void desalocarLista(Lista * lista) {
    Celula * c, * proxADesalocar;
    for (c = lista->primeiro; c != NULL; c = proxADesalocar) {
        proxADesalocar = c->prox;
        free(c);
    }
    proxADesalocar = NULL;
    c = NULL;
}
